File: core/ConstructStringFromBinaryTree/ConstructStringFromBinaryTree.py
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def tree2str(root: Optional[TreeNode]) -> str:
    def dfs(node):
        if node is None:
            return ""
        res = f"{node.val}"
        left = dfs(node.left)
        right = dfs(node.right)
        if left == "" and right == "":
            return res
        elif right == "":
            return f"{res}({left})"
        else:
            return f"{res}({left})({right})"

    return dfs(root)

File: core/ConstructStringFromBinaryTree/ConstructStringFromBinaryTree.java
package com.dbc;

public class ConstructStringFromBinaryTree {
    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode() {
        }

        TreeNode(int val) {
            this.val = val;
        }

        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }

    public String tree2str(TreeNode root) {
        if (root == null) {
            return "";
        }
        String res = "" + root.val;
        String left = tree2str(root.left);
        String right = tree2str(root.right);
        if (left.equals("") && right.equals("")) {
            return res;
        } else if (right.equals("")) {
            return res + "(" + left + ")";
        } else {
            return res + "(" + left + ")" + "(" + right + ")";
        }
    }
}
